doc/source/devel/array.rst

```diff
@@ -0,0 +1,612 @@
+
+
+    * Construct an array of shape ``(5,)`` that holds ints:
+       
+    .. code-block:: c++
+
+       CelloArray<int,1> arr(5); 
+       CelloArray<int,1> arr2 = CelloArray<int,1>(5); // yields same result
+
+Wrap a pre-existing pointer:
+
+  * Use the constructor ``CelloArray(T* array, Args... args);`` to wrap the
+    pointer ``array`` which represents an array with shape
+    ``(arg0, arg1, ... arg{D-1})``.
+
+  * Example: Construct an array representing ``[[0,1,2],[3,4,5]]``:
+
+  .. code-block:: c++
+
+     int data[] = {0,1,2,3,4,5};
+     CelloArray<int,2> arr(data,2,3);
+
+We can also forward declare an array and assign values to it later.
+
+.. code-block:: c++
+
+   int data[] = {0,1,2,3,4,5};
+   CelloArray<int,2> arr; 
+   arr = CelloArray<int,2>(data,2,3);
+
+
+Dimension Size
+--------------
+
+To get the length along a dimension (or axis), call
+``arr.shape(unsigned int dim)``, where ``dim`` is the number of the
+dimension. Dimensions numbers start at ``0`` and are ordered with
+increasing indexing speed (``dim=D-1`` is the dimension with fastest
+indexing).
+
+Element Access
+--------------
+
+To access an element pass indices to the ``operator()(Args... args)``
+method. As many indices should be specified as there are dimensions in
+the array (the number of args **must** match the number of dimensions.
+
+The ``operator()(Args... args)`` method returns a reference or copy
+(depending on the circumstance) of the element.
+
+**Example:** print element ``(0,2)`` of the array ``[[0,1,2],[3,4,5]]``:
+
+.. code-block:: c++
+
+   int data[] = {0,1,2,3,4,5};
+   CelloArray<int,2> arr(data,2,3);
+   printf("%d\n", arr(0,2)); // prints "2"
+   // printf("%d\n", arr(2));       This would fail to compile
+   // printf("%d\n", arr(0,0,2));   This would fail to compile
+
+
+Simple Assignment - Shallow/Deep Copies
+---------------------------------------
+
+Shallow copies are produced via ordinary assignment.
+
+.. code-block:: c++
+
+   int data[] = {0,1,2,3,4,5};
+   CelloArray<int,2> a(data,2,3);
+   CelloArray<int,2> b = a; // b is now a shallow copy of arr
+   CelloArray<int,2> c(2,2); // c represents [[0,0],[0,0]]
+   CelloArray<int,2> d = c; // d is now a shallow copy of c
+   c = a; // c is now a shallow copy of a
+
+When ``c`` is assinged the contents of ``a``, ``c`` becomes a shallow
+copy of ``a``. However the contents of ``d`` are unaffected.  It still
+represents the array ``[[0,0],[0,0]]``.
+
+To perform a deepcopy, assign the the results of the ``deepcopy`` method.
+
+.. code-block:: c++
+
+   int data[] = {0,1,2,3,4,5};
+   CelloArray<int,2> a(data,2,3);
+   CelloArray<int,2> b = a; // b is now a shallow copy of arr
+   CelloArray<int,2> c(2,2); // c represents [[0,0],[0,0]]
+   CelloArray<int,2> d = c; // d is now a shallow copy of c
+   c = a; // c is now a shallow copy of a
+   
+Modifications to the contents of ``e`` will not be reflected in ``a``
+or ``data`` (and vice-versa)
+
+
+Creating Subarrays
```

Comment:
I've actually started to wonder about that myself.  I totally agree that it's worth thinking more about. 

This is somewhat speculative but my instincts tell me that if subarrays are made locally within a function, it should be fairly transparent to the compiler. But, I'm a little worried that the `std::shared_ptr` underlying the `CelloArray` could obfuscate things (I wounder if Kokkos Views handle this better - they definitely do something more sophisticated than simply wrapping `std::shared_ptr` because they don't increment reference counts for subviews created in `KOKKOS_LAMBDA`)

Reply:
I think as long as `std::shared_ptr`s aren't being created during vectorized sections, only accessed, it should be good.